### PR TITLE
ApolloProvider

### PR DESCRIPTION
**Pull request by abhiaiyer91**
After much iteration, the best way to interact with the redux provider is through a wrapper component. We ended up changing the namespace of `Provider` to `ApolloProvider`, and uses `react-redux` provider. Added some more tests too.

### CODE REVIEW

**Comment by abhiaiyer91** [comment]
Man, Test coverage went down lol. wth

**Comment by abhiaiyer91** [comment]
How shall the version bump go?

**Comment by stubailo** [comment]
@abhiaiyer91 sigh, coveralls is really unreliable.

Let's bump the version after merging!

**Comment by abhiaiyer91** [comment]
@stubailo I made a mistake and also grouped the namespace change as well!

**Comment by abhiaiyer91** [comment]
@stubailo @jbaxleyiii sorry for mixing in the namespace change with this body of work. I can reopen 2 PRs if necessary! Let me know!

**Comment by abhiaiyer91** [comment]
We should also rewrite the docs and example apps with new Namespace if merged. And add a section for the composable provider. Would you like an issue opened in the docs or here in `react-apollo`

**Comment by jbaxleyiii** [comment]
The docs would be great and mention this PR

**Comment by stubailo** [comment]
NP, one PR is fine!

**Comment by abhiaiyer91** [comment]
Alright issues made. Whenever you guys get a chance to give some feedback, that'd be great! Once merged @stubailo assign the issue in the docs to me!

File: src/composeApolloProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/// <reference path="../typings/main.d.ts" />
+/* tslint:disable:no-unused-variable */
```
Comment (abhiaiyer91) [inline src/composeApolloProvider.tsx]: this should be removed.

**Comment by jbaxleyiii** [comment]
👍 I'll take a look tonight

**Comment by abhiaiyer91** [comment]
@jbaxleyiii @stubailo show this PR some love! haha

File: src/Provider.tsx [hunk 2]
```diff
@@ -17,7 +17,7 @@ export declare interface ProviderProps {
   client: ApolloClient;
 }
 
-export default class Provider extends Component<ProviderProps, any> {
+export default class ApolloProvider extends Component<ProviderProps, any> {
```
Comment (stubailo) [inline src/Provider.tsx]: Rename the file, as well

Comment (stubailo) [inline src/Provider.tsx]: Also, shouldn't `ApolloProvider` just use `composeApolloProvider` under the hood now?

File: src/composeApolloProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+declare interface ProviderProps {
+    store?: Store<any>;
+    client: ApolloClient;
+}
+
+export default function composeApolloProvider<T>(Provider: React.ComponentClass<any>): any {
+    return class WrapperProvider extends Component<ProviderProps, any> {
+        /* tslint:disable:member-ordering */
+        static propTypes = {
+            store: PropTypes.shape({
```
Comment (stubailo) [inline src/composeApolloProvider.tsx]: Any way we can extract this from `Provider` itself, or just pass it through?

File: src/composeApolloProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+
+import ApolloClient from 'apollo-client';
+
+declare interface ProviderProps {
+    store?: Store<any>;
+    client: ApolloClient;
+}
+
+export default function composeApolloProvider<T>(Provider: React.ComponentClass<any>): any {
```
Comment (stubailo) [inline src/composeApolloProvider.tsx]: Is this function better than simply having `ApolloProvider` always use `Provider` as the implementation? As long as it's a peer dependency, the developer can choose whatever version of the Redux provider they feel like.

Comment (stubailo) [inline src/composeApolloProvider.tsx]: I guess I feel like it just adds an extra line of code, but doesn't really add much clarity since you still don't know what it's doing with the `Provider`. If we wanted people to have the ultimate clarity about what the provider is doing we could just ask them to use two providers:

```js
<Provider>
   <ApolloProvider>
      ...
```

File: src/Provider.tsx (hunk 2 above)
Comment (abhiaiyer91) [inline src/Provider.tsx]: Yeah those are the kinda questions I wanted to talk through. Since we added some duplication with this function. 

I think `ApolloProvider` should be shipped using `composeApolloProvider` under the hood.

File: src/composeApolloProvider.tsx (hunk 3 above)
Comment (abhiaiyer91) [inline src/composeApolloProvider.tsx]: If we change the underlying implementation of Provider this duplication is cleared

File: src/composeApolloProvider.tsx (hunk 4 above)
Comment (abhiaiyer91) [inline src/composeApolloProvider.tsx]: Actually I think now that we are actually using Redux under the hood for the provider, we should just recommend people use this directly. Now that my feet are more wet into Apollo, I think using the ApolloProvider directly with this implementation makes more sense, than composing it yourself in app.

**Comment by abhiaiyer91** [comment]
Thanks for the feedback, will iterate again!

**Comment by stubailo** [comment]
Thanks - I'm going to try to get you that architectural overview ASAP

**Comment by abhiaiyer91** [comment]
@stubailo @jbaxleyiii i think this is good to go!

File: test/ApolloProivder.tsx [hunk 5]
```diff
@@ -0,0 +1,97 @@
+/// <reference path="../typings/main.d.ts" />
```
Comment (stubailo) [inline test/ApolloProivder.tsx]: Typo in this file name!

**Comment by jbaxleyiii** [comment]
@abhiaiyer91 @stubailo I'm good with this! We will need to update docs and the Changelog. I'll publish it as v0.3.0 after we merge

**Comment by stubailo** [comment]
I think we need a corresponding docs PR before merging.

**Comment by stubailo** [comment]
Also, this PR changes everything from 2-space tabs to 4-space tabs, which isn't great.

**Comment by jbaxleyiii** [comment]
I didn't catch that. I'm surprised the linter didnt bug out. Let's fix that

**Comment by abhiaiyer91** [comment]
I just noticed. WebStorm messing with me...

**Comment by abhiaiyer91** [comment]
Let me fix!

File: test/ApolloProivder.tsx (hunk 5 above)
Comment (abhiaiyer91) [inline test/ApolloProivder.tsx]: shit man sorry!

**Comment by stubailo** [comment]
I think about a year ago all of the linters collectively gave up on linting whitespace since there are so many edge cases. but maybe it's possible.

I don't think any linter can save us from inspecting the code :]

**Comment by abhiaiyer91** [comment]
Alright those problems are addressed. Whats next?

File: src/ApolloProvider.tsx [hunk 6]
```diff
@@ -33,8 +38,8 @@ export default class Provider extends Component<ProviderProps, any> {
     client: PropTypes.object.isRequired,
   };
 
-  public store: Store<any>;
-  public client: ApolloClient;
+  public store:Store<any>;
```
Comment (stubailo) [inline src/ApolloProvider.tsx]: This should have a space, I thought TSLint enforced this. @abhiaiyer91 you might want to set up TSLint in your editor.

**Comment by abhiaiyer91** [comment]
Alright Editor is all setup! Hahah my bad. I hit the "Typescript" button in WebStorm, turns out my settings were F'd up.

**Comment by stubailo** [comment]
Feel free to commit in a WebStorm config! We have one for vscode in the repo.

**Comment by stubailo** [comment]
Looks great to me!

Last step is the docs PR, then we can merge and publish!

**Comment by abhiaiyer91** [comment]
Okay I'll get the docs PR together now.

**Comment by abhiaiyer91** [comment]
Docs PR https://github.com/apollostack/docs/pull/61

**Comment by stubailo** [comment]
OK, up to @jbaxleyiii to merge and publish this and the docs change at the same time!
